### ISSUE TITLE
add support for crypttab tmp= mode

### ISSUE THREAD

**Issue by TheJJ** [systemd/systemd#952]
https://github.com/systemd/systemd/blob/master/src/cryptsetup/cryptsetup.c#L203

systemd does not handle the `tmp=btrfs` option (as noted in `man crypttab`) to create a specific filesystem on a crypttab device on the fly.

Plain `tmp` does work, just the filesystem option is not supported.

**Comment by poettering** [comment]
Hmm? Is that a recent addition to Debian's crypttab script or so?

systemd ships crypttab(5) and that certainly mentions no tmp=ext4 syntax, hence Debian seems to replace this? @mbiebl what's going in?

In general, I am fine with extending the tmp crypttab option like this, would need a patch though...

**Comment by mbiebl** [comment]
See http://linux.die.net/man/5/crypttab

```
tmp
     The encrypted block device will be prepared for using it as tmp partition: it will be formatted using 
     mke2fs and its root directory will be set to mode 01777. The warning about the swap option applies
     here as well.
```

The Debian crypttab man page contains the following

```
tmp=<tmpfs>
        Run mkfs with filesystem type <tmpfs> on the created device. Default is
        ext4.
```

**Comment by poettering** [comment]
The die.net man page is which one precisley? doesn't look like systemd's upstream one, nor is debian's?

@mbiebl, was the `tmp=` syntax always available on Debian?

**Comment by mbiebl** [comment]
@poettering right, the debian cryptsetup man page looks different. I don't want to copy it in full here, but http://manpages.ubuntu.com/manpages/vivid/man5/crypttab.5.html looks like it's pretty close to what it is in Debian.

As far as tmp= goes, I found this in the debian changelog:

cryptsetup (2:1.0.6~pre1+svn45-1) unstable; urgency=low
- Add support for a custom filesystem for /tmp. Patch provided by
  Hans-Peter Oeri hp@oeri.ch.
  
  -- Jonas Meurer mejo@debian.org  Wed, 30 Jan 2008 09:01:52 +0100

I'm not the maintainer of cryptsetup though, so take this information with a grain of salt.

**Comment by poettering** [comment]
OK. Sounds fine to me to add a debian-compatible extension to our code, that maps tmp=<foo> to calling /bin/mkfs.<foo> on the block device.

**Comment by mbiebl** [comment]
Linking the downstream bug for completeness sake:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=773157

**Comment by tomh0665** [comment]
In fairness to Debian, tmp is documented in the crypttab(5) man page of systemd-208-9.fc20 and initscripts-9.20.1-1.fc14. So it must've been available in Fedora pre-systemd.

**Comment by poettering** [comment]
@tomh0665 the "tmp" syntax is supported and documented by systemd's cryptsetup, but the "tmp=ext4" syntax is neither supported nor documented. I think you must be misreading the man page there.

**Comment by poettering** [comment]
Fix waiting in #15853
